Populate Main Branch with Initial Code

Gitmoid opened the pull request:
This pull request introduces the initial codebase to the main branch. The main branch was intentionally emptied to facilitate a clean import and review of the code from the code-review branch.

Gitmoid committed:
add all files to code-review manually
unable to create a pull request otherwise

jankoscak commented on pom.xml:
```diff
@@ -0,0 +1,90 @@
+	<version>0.0.1-SNAPSHOT</version>
+	<name>demo</name>
+	<description>Demo project for Spring Boot</description>
+	<url/>
+	<licenses>
+		<license/>
+	</licenses>
+	<developers>
+		<developer/>
+	</developers>
+	<scm>
+		<connection/>
+		<developerConnection/>
+		<tag/>
+		<url/>
+	</scm>
```
:art: Zbytecne kdyz to neni vyplnene.

jankoscak commented on src/main/java/com/vrana/database/config/MapperConfig.java:
```diff
@@ -0,0 +1,23 @@
+    }
+}
+
+// guess it handles nested authors automatically now
+//    @Bean
+//    public ModelMapper modelMapper() {
+//        ModelMapper modelMapper = new ModelMapper();
+//        modelMapper.getConfiguration().setMatchingStrategy(MatchingStrategies.LOOSE);
+//        return modelMapper;
```
🐛 Pryc so zakomentovanym kodem.

jankoscak commented on src/main/java/com/vrana/database/BookApiApplication.java:
```diff
@@ -0,0 +1,31 @@
+		SpringApplication.run(BookApiApplication.class, args);
+	}
+
+	@Override
+	public void run(String... args) throws Exception {
+		log.info("Datasource: " + dataSource.toString());
+		final JdbcTemplate restTemplate = new JdbcTemplate(dataSource);
+		restTemplate.execute("select 1");
+	}
```
🐛 Tohlecto neni potreba.

jankoscak commented on src/main/java/com/vrana/database/services/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        bookEntity.setIsbn(isbn);
+
+        return bookRepository.findById(isbn).map(existingBook -> {
+            Optional.ofNullable(bookEntity.getTitle()).ifPresent(existingBook::setTitle);
+            return bookRepository.save(existingBook);
+        }).orElseThrow(() -> new RuntimeException("Book does not exist"));
+    }
```
🐛 ❌   `Optional` a nejakou konkretni vyjimku.

jankoscak commented on src/main/java/com/vrana/database/services/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public List<BookEntity> findAll() {
+        return StreamSupport.stream(bookRepository
+                .findAll()
+                .spliterator(), false)
+                .collect(Collectors.toList());
```
🐛 Ak si odmyslime, ze ziskani vsech prvku bez strankovani neni dobrej napad. (Muze jich byt fakt hodne), tak pak nevim naco je ten `StreamSupport`

```
return bookRepository.findAll()
```

Gitmoid replied:
Myslím, že to byla první verze `findAll()`, než jsem přidal stránkování. `CrudRepository` vrací `Iterable`, proto `StreamSupport`, aby vrátil `List`. Teoreticky to můžu odstranit, nikde to nepoužívám.

jankoscak commented on src/main/java/com/vrana/database/services/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public BookEntity createUpdateBook(String isbn, BookEntity book) {
+        book.setIsbn(isbn);
```
🐛 Upravil bych mapper, aby se to isbn nastavovalo pri vytvareni instance BookEntity. Tady to hrozi bugem.

jankoscak commented on src/main/java/com/vrana/database/services/impl/AuthorServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public boolean isExists(Long id) {
```
:art: `isExists` -> `exists`

jankoscak commented on src/main/java/com/vrana/database/repositories/AuthorRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+    Iterable<AuthorEntity> ageLessThan(int age);
+
+    @Query("SELECT a from AuthorEntity a where a.age > ?1") // HQL annotation is needed
```
🐛 `Query` je tady nadbytecne.

jankoscak commented on src/main/java/com/vrana/database/services/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+
+    private BookRepository bookRepository;
+
+    public BookServiceImpl(BookRepository bookRepository) {
+        this.bookRepository = bookRepository;
+    }
```
🐛 Je mozne pouzit lombok.

jankoscak commented on src/main/java/com/vrana/database/domain/entities/BookEntity.java:
```diff
@@ -0,0 +1,25 @@
+package com.vrana.database.domain.entities;
+
+import jakarta.persistence.*;
```
🐛 Nepouzivat hvezdickove importy. Naimportujete kopec balastu.

jankoscak commented on src/main/java/com/vrana/database/domain/dto/BookDto.java:
```diff
@@ -0,0 +1,24 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
```
🐛 Je potrebny ten builder? Nikde ho nepouzivate.

Gitmoid replied:
používám ho v TestDataUtil

jankoscak replied:
No napsat metodu do tridy jenom proto, ze ji potrebujete v testoch je spatne.

jankoscak reviewed (changes requested):
Prosim zapracujte pripominky. Komentare jsem psal jenom na jedno misto, kde se vyskytoval nejakej nedostatek. Prosim opravte to vsude.

jankoscak commented:
@Gitmoid Upravy, ktere jsem tady pozadoval jste jeste nezpracoval? Nebo mi neco uniklo?

Gitmoid commented:
@jankoscak pár dní po Velikonocích jsem nebyl v Brně a neměl jsem moc času, dnes dodělávám testy ve spocku (chvilku mi trvalo, než jsem přišel na to, jak to správně nastavit s flyway). Na změny se pustím hned potom. Děkuji moc za strpení.

Gitmoid committed:
fixed: remove empty pom.xml fields

Gitmoid committed:
fixed: remove commented-out code

Gitmoid committed:
fixed: remove unused DataSource

Gitmoid committed:
fixed: isExists -> exists

Gitmoid committed:
fixed: remove unnecessary @Query

Gitmoid committed:
fixed: lombok

Gitmoid committed:
fixed: remove unnecessary imports

Gitmoid committed:
fixed: isExists -> exists

Gitmoid committed:
fixed: set isbn in mapper

Gitmoid committed:
fixed: SpringBootTest -> DataJpaTest

Gitmoid committed:
fixed: remove `Optional` and provide a more specific exception

Gitmoid committed:
no exception is needed since `BookController` already checks if a book with this isbn exists

Gitmoid committed:
replace `StreamSupport` with `ArrayList`

jankoscak commented on src/main/java/com/vrana/database/BookApiApplication.java:
```diff
@@ -0,0 +1,19 @@
+
+@SpringBootApplication
+@Log
+public class BookApiApplication implements CommandLineRunner {
```
:bug: Proc je potreba `CommandLineRunner`? Je to webova aplikace.

jankoscak commented on src/main/java/com/vrana/database/services/BookService.java:
```diff
@@ -0,0 +1,26 @@
+
+public interface BookService {
+
+    BookEntity createUpdateBook(BookEntity book);
```
:bug: Metoda by mela mit jenom jeden ucel. Dela update a nebo create? Pro update muze byt jina logika jak pro create. Kdyz je stejna, tak nebylo by vhodnejsi `saveBook`?

jankoscak commented on src/main/java/com/vrana/database/services/BookService.java:
```diff
@@ -0,0 +1,26 @@
+
+    boolean exists(String isbn);
+
+    BookEntity partialUpdate(BookEntity bookEntity);
```
:bug: `partialUpdate` a pritom je ako parameter ocekavan celej objekt. Co se aktualizuje? Aspon kdyby byl javadoc, ktery to specifikuje.

jankoscak commented on src/main/java/com/vrana/database/services/impl/AuthorServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+        authorEntity.setId(id);
+
+        return authorRepository.findById(id).map(existingAuthor -> {
+            Optional.ofNullable(authorEntity.getName()).ifPresent(existingAuthor::setName);
+            Optional.ofNullable(authorEntity.getAge()).ifPresent(existingAuthor::setAge);
```
:bug: Tady jste to neopravil.

jankoscak commented:
@Gitmoid na ostatne pull requesty se nepodivam dokud nedokoncime tohlecto.

Gitmoid committed:
move mapper operations from controller to service layer
add MapStruct for full and partial update
change service methods to accept and return Dto instead of Entity
separate book create and update methods

Gitmoid committed:
add required dependencies
add validation and mapping for author
separate book `create` and `update` method in controller

Gitmoid committed:
fixed: remove `CommandLineRunner`

Gitmoid committed:
fixed: remove `@Builder`

Gitmoid committed:
fix tests
add test containers
handle exceptions to return `ResponseEntity`
remove excessive dependencies
remove wildcard imports
